base/array.jl

```diff
@@ -2327,6 +2327,30 @@ julia> filter(isodd, a)
 """
 filter(f, As::AbstractArray) = As[map(f, As)::AbstractArray{Bool}]
 
+function filter(f, a::Vector{T}) where T
+    j = 1
+    b = Vector{T}(undef, length(a))
+    for ai in a
+        @inbounds b[j] = ai
+        j = ifelse(f(ai), j+1, j)
+    end
+    resize!(b, j-1)
+    sizehint!(b, length(b))
+    b
+end
+
+function filter(f, a::AbstractVector)
+    j=1
+    offset = Int(firstindex(a))-1
+    idxs = Vector{Int}(undef, length(a))
+    for ai in a
+        @inbounds idxs[j] = j + offset
+        j = ifelse(f(ai), j+1, j)
+    end
+    resize!(idxs, j-1)
+    return a[idxs]
```

Comment:
I don't think this is generally the best algorithm. In some cases the extra index vector might be faster, but not enough to justify using twice the memory in every case.

Reply:
Well, it is using up to 8x the temporary memory (`Int` instead of `Bool`), same as a `view`.

However, the temp is very temporary. Should we manually free its buffer immediately via `resize!(idxs, 0); sizehint!(idxs, 0)` in order to limit gc pressure? I am not seeing lots of precedent of that in Base, but I like the idea.

Or is the issue that out-of-place filtering of giant vectors of tiny objects can cause oom errors?